staging/src/k8s.io/kms/encryption/service.go

```diff
@@ -0,0 +1,249 @@
+	transformer, resp, err := m.getTransformerForEncryption(uid)
+	if err != nil {
+		klog.V(2).InfoS("encrypt plaintext", "id", uid, "err", err)
+		return nil, err
+	}
+
+	ct, err := transformer.TransformToStorage(ctx, pt, emptyContext)
+	if err != nil {
+		klog.V(2).InfoS("encrypt plaintext", "id", uid, "err", err)
+		return nil, err
+	}
+
+	return &service.EncryptResponse{
+		Ciphertext:  ct,
+		KeyID:       resp.KeyID, // TODO what about rotation ??
+		Annotations: resp.Annotations,
+	}, nil
+}
+
+func (m *LocalKEKService) getTransformerForDecryption(ctx context.Context, uid string, req *service.DecryptRequest) (value.Transformer, error) {
+	encKEK := req.Annotations[referenceKEKAnnotationKey]
+
+	if _transformer, found := m.transformers.Get(encKEK); found {
+		return _transformer.(value.Transformer), nil
+	}
+
+	key, err := m.remoteKMS.Decrypt(ctx, uid, &service.DecryptRequest{
+		Ciphertext:  encKEK,
+		KeyID:       req.KeyID,
+		Annotations: annotationsWithoutReferenceKeys(req.Annotations),
+	})
+	if err != nil {
+		return nil, err
+	}
+
+	block, err := aes.NewCipher(key)
+	if err != nil {
+		return nil, err
+	}
+	transformer := m.baseTransformerFunc(block)
+
+	// Overwrite the plain key with 0s.
+	copy(key, make([]byte, len(key)))
+
+	m.transformers.Add(encKEK, transformer)
+
+	return transformer, nil
+}
+
+// Decrypt attempts to decrypt the ciphertext with the localKEK, a KEK from the
+// store, or the remote KMS.
+func (m *LocalKEKService) Decrypt(ctx context.Context, uid string, req *service.DecryptRequest) ([]byte, error) {
+	if _, ok := req.Annotations[referenceKEKAnnotationKey]; !ok {
+		return nil, fmt.Errorf("unable to find local KEK for request with uid %q", uid)
+	}
+
+	transformer, err := m.getTransformerForDecryption(ctx, uid, req)
+	if err != nil {
+		klog.V(2).InfoS("decrypt ciphertext", "id", uid, "err", err)
+		return nil, fmt.Errorf("failed to get transformer for decryption: %w", err)
+	}
+
+	pt, _, err := transformer.TransformFromStorage(ctx, req.Ciphertext, emptyContext)
+	if err != nil {
+		klog.V(2).InfoS("decrypt ciphertext with pulled key", "id", uid, "err", err)
+		return nil, err
+	}
+
+	return pt, nil
+}
+
+// Status returns the status of the remote KMS.
+func (m *LocalKEKService) Status(ctx context.Context) (*service.StatusResponse, error) {
+	// TODO(aramase): the response from the remote KMS is funneled through without any validation/action.
+	// This needs to handle the case when remote KEK has changed. The local KEK needs to be rotated and
+	// re-encrypted with the new remote KEK.
+	return m.remoteKMS.Status(ctx)
+}
+
+func annotationsWithoutReferenceKeys(annotations map[string][]byte) map[string][]byte {
+	if len(annotations) <= numAnnotations {
+		return nil
+	}
+
+	m := make(map[string][]byte, len(annotations)-numAnnotations)
+	for k, v := range annotations {
+		if strings.HasSuffix(k, referenceSuffix) {
```

Comment:
Probably does not matter here but I prefer to be extra careful when the loop variable escapes the `for` loop.

```suggestion
	for k, v := range annotations {
		k, v := k, v
		if strings.HasSuffix(k, referenceSuffix) {
```